#!/usr/bin/env python3
PACKAGE = "opencv_final"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

# ========== 颜色阈值参数 ==========
# 红色阈值
gen.add("red_hue_min", int_t, 0, "红色Hue最小值", 0, 0, 180)
gen.add("red_hue_max", int_t, 0, "红色Hue最大值", 10, 0, 180)
gen.add("red_sat_min", int_t, 0, "红色饱和度最小值", 100, 0, 255)
gen.add("red_sat_max", int_t, 0, "红色饱和度最大值", 255, 0, 255)
gen.add("red_val_min", int_t, 0, "红色亮度最小值", 50, 0, 255)
gen.add("red_val_max", int_t, 0, "红色亮度最大值", 255, 0, 255)

# 蓝色阈值
gen.add("blue_hue_min", int_t, 0, "蓝色Hue最小值", 100, 0, 180)
gen.add("blue_hue_max", int_t, 0, "蓝色Hue最大值", 130, 0, 180)
gen.add("blue_sat_min", int_t, 0, "蓝色饱和度最小值", 100, 0, 255)
gen.add("blue_sat_max", int_t, 0, "蓝色饱和度最大值", 255, 0, 255)
gen.add("blue_val_min", int_t, 0, "蓝色亮度最小值", 50, 0, 255)
gen.add("blue_val_max", int_t, 0, "蓝色亮度最大值", 255, 0, 255)

# 敌方颜色
gen.add("enemy_color", int_t, 0, "敌方颜色 (0:红色, 1:蓝色)", 0, 0, 1)

# ========== 灯条参数 ==========
gen.add("light_min_area", double_t, 0, "灯条最小面积", 50.0, 0, 100)
gen.add("light_max_ratio", double_t, 0, "灯条最大宽高比", 0.6, 0, 2)
gen.add("light_min_solidity", double_t, 0, "灯条最小凸度", 0.7, 0, 1)
gen.add("light_min_height", double_t, 0, "灯条最小高度", 14.0, 0, 100)

# ========== 配对参数 ==========
gen.add("light_max_angle_diff", double_t, 0, "灯条最大角度差", 15.0, 0, 90)
gen.add("light_max_height_diff_ratio", double_t, 0, "灯条最大高度差比率", 0.5, 0, 2)
gen.add("light_max_y_diff_ratio", double_t, 0, "灯条最大Y方向差比率", 0.3, 0, 2)
gen.add("light_min_x_diff_ratio", double_t, 0, "灯条最小X方向差比率", 0.5, 0, 2)
gen.add("armor_min_aspect_ratio", double_t, 0, "装甲板最小长宽比", 0.5, 0, 10)
gen.add("armor_max_aspect_ratio", double_t, 0, "装甲板最大长宽比", 4.8, 0, 10)
gen.add("armor_big_armor_ratio", double_t, 0, "大装甲板长宽比阈值", 3.0, 0, 10)
gen.add("armor_small_armor_ratio", double_t, 0, "小装甲板长宽比阈值", 2.0, 0, 10)

# ========== 预处理参数 ==========
gen.add("brightness_thresh", int_t, 0, "亮度阈值", 150, 0, 255)

# ========== 形态学参数 ==========
gen.add("morph_open_size", int_t, 0, "开运算核大小", 3, 0, 20)
gen.add("morph_close_size", int_t, 0, "闭运算核大小", 5, 0, 20)
gen.add("morph_kernel_size", int_t, 0, "形态学核大小", 4, 0, 20)

# ========== 数字识别参数 ==========
gen.add("digit_big", int_t, 0, "大装甲板数字模板大小", 250, 20, 300)
gen.add("digit_small", int_t, 0, "小装甲板数字模板大小", 28, 20, 300)
gen.add("digit_threshold", double_t, 0, "数字模板匹配阈值", 0.75, 0, 1)

# ========== 绘制开关 ==========
gen.add("draw_e_light", bool_t, 0, "绘制灯条", True)
gen.add("draw_e_rect", bool_t, 0, "绘制装甲板矩形", True)
gen.add("draw_e_l_center", bool_t, 0, "绘制灯条中心连线", True)
gen.add("draw_e_r_center", bool_t, 0, "绘制装甲板中心点", True)
gen.add("draw_e_label", bool_t, 0, "绘制标签", True)

# ========== 绘制颜色 ==========
gen.add("draw_color_e_light", int_t, 0, "灯条颜色 (0-7)", 7, 0, 7)
gen.add("draw_color_e_rect", int_t, 0, "装甲板矩形颜色 (0-7)", 5, 0, 7)
gen.add("draw_color_e_l_center", int_t, 0, "灯条中心连线颜色 (0-7)", 4, 0, 7)
gen.add("draw_color_e_r_center", int_t, 0, "装甲板中心点颜色 (0-7)", 3, 0, 7)

exit(gen.generate(PACKAGE, "armor_detect", "armorDetect"))